### ISSUE TITLE
New use of go:embed prevents use of boilingcore as library

### ISSUE THREAD

**Issue by zikaeroh** [volatiletech/sqlboiler#999]
TL;DR: The latest release removed bindata in favor of `go:embed` (213f39685322dbd79e4f188eeedeba0d614ae72f), but in doing so prevents use of the generation code as a library.

-- 

I run sqlboiler's generation code directly so I can avoid having platform-specific scripts or external dependencies: https://github.com/hortbot/hortbot/blob/master/internal/cmd/genmodels/main.go

The change made to `boilingcore` is not backwards/SIV compatible, as it changed the signature of `New`; that's not good itself, but could be worked around. Unfortunately, the FS that is needed to make things work is defined in `main.go`, which isn't accessible (as it's both the main package and unexported). These two together basically mean that nobody can use SQLBoiler's generation code as a library anymore in their own code.

I'm thinking the pattern chosen was only chosen because `templates` and `templates_test` are root directories in the repo. The function itself is only used twice in the project, once in the `main.go` with the embedded data, and once in the tests with `os.DirFS` to access a parent directory.

Can these two folders be placed in `boilingcore` instead, then just have a single embbedded FS that's used, restoring `New` to its old parameter?

I'm more than happy to submit a PR to do this.

**Comment by aarondl** [comment]
Hi @zikaeroh. Sorry I've been slow to fix this issue. I'm hoping to resolve it tonight.

This pattern was chosen because a library should not bundle compile time assets (IMHO). The better directory structure was just a bonus.

Unfortunately it does seem that there's no way to preserve your existing use of sqlboiler as a library and also do the right thing. I think what we'll have to do is revert in a 4.7.1 as a bug fix, and then release v5 with the correct API and behavior.

**Comment by zikaeroh** [comment]
I guess I don't see why a library wouldn't be allowed to include assets. If you don't want all users of boilingcore to have this by default (and I'm not sure why, it's not useful without them, not that it has many importers as it's just the generation side), then can the templates at least be moved to some other directory that exports an FS that could be passed to it? Then, it doesn't include them by default, and others can choose to grab them as needed.

It'd be a bit unfortunate to have to be stuck at 4.7.1 forever, lest I have to bring back shell scripts and wrappers to run the generator. The method I use is a lot more convenient; no need to install any binaries or anything.

**Comment by zikaeroh** [comment]
Maybe I'm misunderstanding the change you're proposing for a different API.

**Comment by aarondl** [comment]
It working like this is mostly an accident of history. The intent is that sqlboiler as a library is useful as something that can generate from any set of templates, that's what the entire boilingcore package is all about, as a library consumer who did not intend to use the default set of templates why would you want all that bloat in your resulting binary? The API is also nicer in that it allows you to override with whatever FS you'd like rather than having to rely on an actual file system easily thanks to the new fs.FS interface.

I think a reasonable compromise is to go back to having a "bindata" package whose sole purpose is to embed it's subdirectories.

```
templates/
  embed.go
  main/
  test/
```

The `templates.Builtin` global variable will include both of these subdirectories.

This will be a package that could be imported by your binary as well as sqlboiler's. In the meantime to preserve backwards compatibility we'll revert the change to `New()` and allow this to be a default dictated by the boilingcore package (which is a bit unfortunate, but we can't throw out semver now can we!?). In v5 we'll break `New()` to accept the argument to the more desirable interface.

Would this work?

**Comment by aarondl** [comment]
Also keep in mind there's a big cost in terms of git history with moving the template files to any directory as we lose the direct history for them and have to dance with git in order to see what happened in the past. I can't say that there was no desire to prevent the move of them. It's a shame we'll end up doing so regardless it seems.

**Comment by zikaeroh** [comment]
I'm personally not too concerned with `boilingcore` being broken API-wise; `pkg.go.dev` only claims two non-fork users, me, and https://github.com/frankie-seb/sinatra/blob/main/plugins/sqlboiler/sqlboiler.go: https://pkg.go.dev/github.com/volatiletech/sqlboiler/v4/boilingcore?tab=importedby

Sourcegraph shows fewer: https://sourcegraph.com/search?q=context:global+github.com/volatiletech/sqlboiler/v4/boilingcore&patternType=literal

The API has _technically_ broken before, most recently a new config item had a different behavior when it was unset (and had to be updated to the old default). See: https://github.com/hortbot/hortbot/commit/4970983637be0964b46759b7b334108af7e6ffc0#diff-1d55c51ab479de48aeede31e0c21ef967e6e976cb6cda883ecc64f4d40c4ce19R79

But, of the two users of `boilingcore` linked above, both use the default templates. I'm not sure that anyone has used it in the way described (for other templates), so I would more likely say to just change the API and not bump the version with so few users. Reasonably, you could also move the templates to a specific package, keep the old `New` call that uses it, then introduce a new function that takes the FS; if old function (and therefore templates) are unused, they'd be dead-code-eliminated during the compilation anyway.

As for the templates, a straight move shouldn't break the history, I believe, but maybe I'm wrong and there's something I'm not thinking of. An alternative would be for me to replicate a FS by instead querying the `go` tool to get the module cache folder for `sqlboiler`, then reference the templates dir there, but that won't work for that sinatra project, and poses another problem where there's nothing to limit the folders that are searched for templates (before, bindata only ran on certain files, and currently `go:embed` does the same).

I'm a lot less concerned about bloat in general, though; all current users of this are CLI tools that run once, and I would also imagine that given the way `boilingcore` works (you must write to a directory), that it's much less likely to be used in contexts that aren't CLI tools. The templates themselves are only around 280 KB.

**Comment by aarondl** [comment]
@zikaeroh I've implemented the changes in the `dev` branch. If you would be so kind to check them over and ensure they meet your expectations I'll get the patch release out and we can put this messiness behind us. Sorry for the problems this release caused.

**Comment by zikaeroh** [comment]
Yes, it seems to work. Thanks!

Note that there's a `fmt.Println` leftover, probably leftover from debugging: https://github.com/volatiletech/sqlboiler/commit/eb0cb08f95c17cf615ea94ed2b676b0de3aead5a#diff-ba374bfc208c800ce9e3786a661ca2e902d105895b74b5268a3a0d543c441a1dR250

**Comment by zikaeroh** [comment]
Checking the changes, this one might be backwards incompatible: https://github.com/volatiletech/sqlboiler/commit/eb0cb08f95c17cf615ea94ed2b676b0de3aead5a#diff-91f66f6d4438e45825670dc2b3499383a14db960677982bdb8a3d5e8af9fc361R133

It may be the case that there are template dirs with `_test` to indicate that they are tests, but now they have to be named "test" instead.

Perhaps you didn't see #1000, but this was the original fix I sent for all of this, which just moved things into boilingcore instead (and kept the names as they were).

**Comment by aarondl** [comment]
Ah good catches. Sorry been doing all of this in a hurry. I will leave both "test" and suffix "_test" as workable. The surface area of that breakage is small enough that I'm not inclined to work around it.

I just wanted to avoid templates/templates and templates/templates_test, there's an annoying amount of stuttering going on there, I also wanted to avoid moving the templates into boilingcore as in your PR since they don't belong in there as I previously mentioned - this sets us up better for a proper `New()` in v5 and a boilingcore package that doesn't understand embedded templates again. Though I'll admit you did all the heavy lifting for me if I had done even a cursory glance at it I would have saved myself some bugs so thanks for that :) I'll try to get this out tonight.

**Comment by zikaeroh** [comment]
All is well with v4.7.1 (which I've happily updated to). Thanks!
